### PR TITLE
Add digest to valid urls in assets

### PR DESCRIPTION
**Pull request by brenogazzola**
Propshaft will not have `asset_path`. Instead we will use `url` so that we can automatically handle external libraries, like bootstrap.

Todo:
- [x] Support all use cases demonstrated in [MDN](https://developer.mozilla.org/en-US/docs/Web/API/URL);
- [x] Automatically adjust relative paths (`../../`, etc.);
- [x] Ensure it's working with real world libraries

### CODE REVIEW

File: test/propshaft/compilers/css_asset_urls_test.rb [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    assert_match /{ background: url\("\/assets\/foobar\/source\/images\/file-[a-z0-9]{40}.jpg"\); }/, compiled
+  end
+
+  test "parent directory" do
+    compiled = compile_asset_with_content(%({ background: url('../file.jpg'); }))
+    assert_match /{ background: url\("\/assets\/foobar\/file-[a-z0-9]{40}.jpg"\); }/, compiled
+  end
+
+  test "grandparent directory" do
+    compiled = compile_asset_with_content(%({ background: url('../../file.jpg'); }))
```
Comment (brenogazzola) [inline test/propshaft/compilers/css_asset_urls_test.rb:59]: Broken. Capture group is only returning a single occurence of `../` in `match[1]`

File: test/propshaft/compilers/css_asset_urls_test.rb [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    assert_match /{ content: url\("\/assets\/foobar\/source\/file-[a-z0-9]{40}.svg"\) url\("\/assets\/foobar\/source\/file-[a-z0-9]{40}.svg"\); }/, compiled
+  end
+
+  test "multiple" do
+    compiled = compile_asset_with_content(%({ content: url(file.svg) url(file.svg); }))
+    assert_match /{ content: url\("\/assets\/foobar\/source\/file-[a-z0-9]{40}.svg"\) url\("\/assets\/foobar\/source\/file-[a-z0-9]{40}.svg"\); }/, compiled
+  end
+
+  test "with svg path" do
+    compiled = compile_asset_with_content(%({ mask-image: url("file.svg#mask1"); }))
```
Comment (brenogazzola) [inline test/propshaft/compilers/css_asset_urls_test.rb:79]: Broken, regex can't separate #mask1 yet.

File: test/propshaft/compilers/css_asset_urls_test.rb [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    assert_match /{ background: url\("\/assets\/foobar\/file-[a-z0-9]{40}.jpg"\); }/, compiled
+  end
+
+  test "grandparent directory" do
+    compiled = compile_asset_with_content(%({ background: url('../../file.jpg'); }))
+    assert_match /{ background: url\("\/assets\/file-[a-z0-9]{40}.jpg"\); }/, compiled
+  end
+
+  test "sibling directory" do
+    compiled = compile_asset_with_content(%({ background: url('../sibling/file.jpg'); }))
```
Comment (brenogazzola) [inline test/propshaft/compilers/css_asset_urls_test.rb:64]: Broken. Regex can't handle this one yet.

**Comment by brenogazzola** [comment]
It works!

**folder structure:**
```
app/assets/vendor/startbootstrap/assets/img/tnw-logo.svg
app/assets/vendor/startbootstrap/assets/css/style.css
```

original css:
```css
.test {
    background-image: url("../assets/img/tnw-logo.svg");
}
```

compiled css:
```css
.test {
    background-image: url("/assets/startbootstrap/assets/img/tnw-logo-[digest].svg");
```

And the image was found and displayed!

**Comment by stevestmartin** [comment]
Thanks @brenogazzola this looks awesome, thanks for doing it so quickly. Looking through the test cases it feels like you've covered everything, I'll see if I can give it a try later this evening.

**Comment by brenogazzola** [comment]
Benching fastest way to grab only the important part of the filename, which means excluding `\s`, `'` and `"`:
https://gist.github.com/brenogazzola/6af4cb65f791bf295aaad6803570359e

Result:
```
Calculating -------------------------------------
            baseline     99.159k (± 2.3%) i/s -      2.979M in  30.055157s
       gsub_wildcard     62.070k (± 2.3%) i/s -      1.867M in  30.094523s
         gsub_except     70.829k (± 2.7%) i/s -      2.128M in  30.064118s
              delete     47.506k (± 4.5%) i/s -      1.422M in  30.015418s
```

Winning version is:
```ruby
input.gsub(/url\(\s*["']?([^"')]+)\s*["']?\)/) do
  "url('#{$1}')"
end
```

**Comment by brenogazzola** [comment]
1 - Inlined the code of the `asset?` method which didn't help much with readability
2 - Simplified the regex. There was not need to separate the pathing
3 - Simplified `resolve_path`.

**Comment by brenogazzola** [comment]
I'm sorely tempted to remove support for the SVG mask. I doubt (but have no evidence) that it's not really used in any of the popular libs and it's imposing a performance penalty in the compiler.

File: lib/propshaft/compilers/css_asset_urls.rb [hunk 4]
```diff
@@ -1,13 +1,40 @@
...
+  def compile(logical_path, input)
+    input.gsub(/url\(\s*["']?([^"'\s)]+)\s*["']?\)/.freeze) do
+      match = Regexp.last_match
+
+      if match[1] =~ /^(http|data|#)/
+        match[0]
+      else
+        resolved_path = resolve_path(logical_path.dirname, match[1])
+        resolved_path, mask = extract_svg_mask(resolved_path.to_s)
+        %[url("#{assembly.config.prefix}/#{assembly.load_path.find(resolved_path).digested_path}#{mask}")]
```
Comment (dmcge) [inline lib/propshaft/compilers/css_asset_urls.rb:19]: What do we want to happen if we can’t find the `resolved_path`? I feel like it should blow up, your CSS shouldn’t compile.

I think it _will_ blow up as written — `find` returns nil and you’ll get a `NoMethodError` for `digested_path`. But that might confuse people and make them think Propshaft is broken. So seems to me that we need either to raise in `find` or define a `find!` method that raises (if there are legit non-raising cases).

**Review by dmcge — COMMENTED** [review]
Nice work ❤️

File: benchmarks/css_asset_urls.rb [hunk 5]
```diff
@@ -0,0 +1,36 @@
+# Benchmark file for the CssAssetUrls compiler
```
Comment (directionless) [inline benchmarks/css_asset_urls.rb:1]: I know this has merged, but I'd been stashing this sort of thing in `tools/benchmarks`

File: lib/propshaft/compilers/css_asset_urls.rb [hunk 6]
```diff
@@ -1,13 +1,34 @@
...
+      directory + filename.delete_prefix("./")
     end
   end
+
+  def extract_svg_mask(resolved_path)
+    mask = resolved_path[/(\#.+)\z/.freeze, 1]
+    path = mask ? resolved_path.sub(mask, "") : resolved_path
+
+    [path, mask]
+  end
```
Comment (directionless) [inline lib/propshaft/compilers/css_asset_urls.rb:33]: I can work through what these do, but I can't verify the correctness. A comment explaining the intent here would help.

File: benchmarks/css_asset_urls.rb (hunk 5 above)
Comment (directionless) [reply benchmarks/css_asset_urls.rb:1]: Er, I commented on the wrong PR. Sorry y'all

**Pull request merged**